### ISSUE TITLE
"Elapsed real-time clock exceeded elapsed monotonic clock" log message

### ISSUE THREAD

**Issue by f2bo** [aws/aws-cdi-sdk#18]
I'm starting to test our application which is sending audio and video streams between two EC2 instances and notice that sometimes the receiver starts to continuously log a large number of messages regarding clocks. I've seen the comments in the code regarding clock drift but I'm not sure I fully understand the significance. They are being logged at INFO level so presumably they are not expected to occur with such frequency. Is this an indication that the application is not handling payloads in a timely fashion and something to be concerned about? I'd appreciate some guidance. 

Thanks

```
...
[avid_in     ] INFO   : 4832 : Elapsed real-time clock[483535224]us exceeded elapsed monotonic clock[483535525]us by [301]us. Using monotonic clock.
[avid_in     ] INFO   : 4832 : Elapsed real-time clock[483536090]us exceeded elapsed monotonic clock[483536390]us by [300]us. Using monotonic clock.
[avid_in     ] INFO   : 4832 : Elapsed real-time clock[483546622]us exceeded elapsed monotonic clock[483546923]us by [301]us. Using monotonic clock.
[avid_in     ] INFO   : 4832 : Elapsed real-time clock[483585742]us exceeded elapsed monotonic clock[483586043]us by [301]us. Using monotonic clock.
[avid_in     ] INFO   : 4832 : Elapsed real-time clock[483586615]us exceeded elapsed monotonic clock[483586916]us by [301]us. Using monotonic clock.
[avid_in     ] INFO   : 4832 : Elapsed real-time clock[483587483]us exceeded elapsed monotonic clock[483587784]us by [301]us. Using monotonic clock.
[avid_in     ] INFO   : 4832 : Elapsed real-time clock[483588359]us exceeded elapsed monotonic clock[483588659]us by [300]us. Using monotonic clock.
[avid_in     ] INFO   : 4832 : Elapsed real-time clock[483595955]us exceeded elapsed monotonic clock[483596255]us by [300]us. Using monotonic clock.
[avid_in     ] INFO   : 4832 : Elapsed real-time clock[483612340]us exceeded elapsed monotonic clock[483612641]us by [301]us. Using monotonic clock.
[avid_in     ] INFO   : 4832 : Elapsed real-time clock[483613207]us exceeded elapsed monotonic clock[483613508]us by [301]us. Using monotonic clock.
[avid_in     ] INFO   : 4832 : Elapsed real-time clock[483624833]us exceeded elapsed monotonic clock[483625134]us by [301]us. Using monotonic clock.
[avid_in     ] INFO   : 4832 : Elapsed real-time clock[483660042]us exceeded elapsed monotonic clock[483660343]us by [301]us. Using monotonic clock.
[avid_in     ] INFO   : 4832 : Elapsed real-time clock[483660904]us exceeded elapsed monotonic clock[483661204]us by [300]us. Using monotonic clock.

...
hundreds more...
```

**Comment by trevorheaws** [comment]
This is an interesting issue. I have not personally encountered that particular message, but I can give a bit more detail about what it is trying to say, in regards to the clock drift message you received. This is checking elapsed time from two different clock sources: the monotonic clock, which is based on the systems start time (and will count at a constant rate) and the realtime clock, which should be synced with some sort of NTP time source and receives adjustments for things like leap seconds. The next leap second will not occur before June 30th 2021 at the earliest so we can rule that out.

It seems then that the likeliest scenario is the realtime clock has undergone some substantial correction (~300 us) on your system. I have a couple of questions:
•	Roughly how long was your test running before you started seeing these messages?
•	Is this repeatable between test runs or was this a one off error?
•	Did the time difference displayed in the message at,  “…. by [301]us …” remain constant or did it change over time?

It is recommended that Amazon Time Sync Service is used with the SDK. Directions for how to configure it on windows can be found at https://docs.aws.amazon.com/AWSEC2/latest/WindowsGuide/windows-set-time.html under the section “To change the NTP server to use the Amazon Time Sync Service”. Let me know if you already configured your instance for using the time sync service.

**Comment by f2bo** [comment]
Hi Trevor,

Some answers:

> Roughly how long was your test running before you started seeing these messages?
- Not long. I usually run quick tests with a 10 min. video and I start seeing these messages soon after it starts. 

> Is this repeatable between test runs or was this a one off error?
- Yes. It's rare that I can play the entire video without seeing any such message. Moreover, once they start, they continue to be emitted very frequently, maybe once per second, and usually continue until the video ends.

> Did the time difference displayed in the message at, “…. by [301]us …” remain constant or did it change over time?
- I wasn't paying particular attention to this, but for the ones that I noticed, it did seem that the times were around that value 301 and 302us. I'll check again.

I set up the EC2 instances some time ago and I don't recall setting up the Amazon Time Sync Service, though I might have. I'll try to do that tomorrow and let you know if I notice any changes.

**Comment by f2bo** [comment]
I checked and the Amazon Time Sync Service was already configured in both machines. 
```
w32tm /query /status
Leap Indicator: 0(no warning)
Stratum: 4 (secondary reference - syncd by (S)NTP)
Precision: -23 (119.209ns per tick)
Root Delay: 0.0003134s
Root Dispersion: 0.8244571s
ReferenceId: 0xA9FEA97B (source IP:  169.254.169.123)
Last Successful Sync Time: 11/26/2020 2:37:12 AM
Source: 169.254.169.123,0x9
Poll Interval: 10 (1024s)
```

I ran a few more tests and this time it seemed that the time difference did not remain constant but slowly incremented by 1us every few seconds starting from around 100us for the first messages, which start almost immediately, and building up to more than 2000us by the end of the video. I'm not sure why this changed because it seemed to be more or less constant previously. 

I finally set `CdiRxConfigData.buffer_delay_ms` to 0 (previously it had been `ENABLED_RX_BUFFER_DELAY_DEFAULT_MS`) and the messages stopped but I believe that's expected. Right?

**Comment by trevorheaws** [comment]
I have so far been unable to recreate the behavior of frequent mismatches between the real-time and monotonic clock. However, there is a change that was merged to mainline as part of the AVM 2.0 release which lowers the priority of recurring errors to DEBUG level messages since there is code in place for handling this condition. There is also now a count value available as part of the connection rx_state to track these events. Additionally, the first occurrence of these clock mismatch events will now generate a WARNING instead of an INFO level message.

**Comment by f2bo** [comment]
Thanks. I'll give it a try when I get the chance to test again with v2.0. 

Regarding the severity of messages, I noticed that v2.0 adds additional logging while waiting for a connection to be established. I can't check right now but it's a couple of messages along the lines of "Waiting for a connection.." and something else that repeat every few seconds and are generated at INFO level. They are a bit noisy, especially if either side is not ready. I wonder if they could be logged with DEBUG level instead.
